File: tests/cpp/src/T043_MultiResponse.cpp
#include "TestCase.h"

#include <vector>

using namespace gmsec::api5;
using namespace gmsec::api5::util;


class T041_Callback : public Callback
{
public:
	T041_Callback(Test& test)
		: m_test(test)
	{
		m_statusList.push_back(Message::ResponseStatus::ACKNOWLEDGEMENT);
		m_statusList.push_back(Message::ResponseStatus::WORKING_KEEP_ALIVE);
		m_statusList.push_back(Message::ResponseStatus::WORKING_KEEP_ALIVE);
		m_statusList.push_back(Message::ResponseStatus::WORKING_KEEP_ALIVE);
		m_statusList.push_back(Message::ResponseStatus::SUCCESSFUL_COMPLETION);
	}

	void onMessage(Connection& conn, const Message& message)
	{
		try
		{
			Message reply = conn.getMessageFactory().createMessage("RESP.DIR");
			reply.setSubject( m_test.getSubject("RESP.DIR").c_str() );

			reply.addField("COMPONENT", (message.hasField("DESTINATION-COMPONENT") ? message.getStringValue("DESTINATION-COMPONENT") : "RESPONDER"));
			reply.addField("DESTINATION-COMPONENT", (message.hasField("COMPONENT") ? message.getStringValue("COMPONENT") : "FILL"));

			if (message.hasField("REQUEST-ID"))
			{
				reply.addField(*message.getField("REQUEST-ID"));
			}

			GMSEC_VERBOSE << "Received Request:\n" << message.toXML();

			for (const auto& status : m_statusList)
			{
				reply.addField("RESPONSE-STATUS", static_cast<GMSEC_I16>(status));

				GMSEC_VERBOSE << "Issuing Reply:\n" << reply.toXML();

				conn.reply(message, reply);
			}
		}
		catch (const GmsecException& e)
		{
			GMSEC_ERROR << "GmsecException in T041_Callback: " << e.what();
		}
	}

	int expectedResponses() const
	{
		return static_cast<int>(m_statusList.size());
	}

private:
	Test& m_test;
	std::vector<Message::ResponseStatus> m_statusList;
};


class T041_ReplyCallback : public ReplyCallback
{
public:
	T041_ReplyCallback()
	{
		repliesReceived = 0;
	}

	void onReply(Connection& conn, const Message& request, const Message& reply)
	{
		try
		{
			GMSEC_INFO << "Received Reply with RESPONSE-STATUS: " << reply.getI16Value("RESPONSE-STATUS");
			++repliesReceived;
		}
		catch (const GmsecException& e)
		{
			GMSEC_ERROR << "GmsecException in T041_ReplyCallback: " << e.what();
		}
	}

	void onEvent(Connection& conn, const Status& status, Connection::Event connEvent)
	{
		GMSEC_WARNING << "Received unexpected event: " << status.get();
	}

	int numRepliesReceived() const
	{
		return repliesReceived;
	}

private:
	int repliesReceived;
};


int test_MultiResponse(Test& test)
{
	test.setDescription("Multi-Response");

	Config config( test.getConfig() );  // make a copy
	config.addValue("mw-multi-resp", "true");

	T041_Callback       cb(test);
	T041_ReplyCallback rcb;

	StdUniquePtr<Connection> conn;

	try
	{
		conn.reset( new Connection(config) );

		conn->connect();

		set_standard_fields( conn->getMessageFactory() );

		conn->startAutoDispatch();

		Message request = conn->getMessageFactory().createMessage("REQ.DIR");
		request.setSubject( test.getSubject("REQ.DIR").c_str() );
		request.addField("DIRECTIVE-STRING", "Do something!");
		request.addField("DESTINATION-COMPONENT", "RESPONDER");
		request.addField("REQUEST-ID", static_cast<GMSEC_U16>(0));

		// subscribe to receive the response message (hack!)
		conn->subscribe( test.getSubject("RESP.DIR.+").c_str() );

		// subscribe to process the request message
		conn->subscribe( request.getSubject(), &cb );

		GMSEC_VERBOSE << "Issuing Request:\n" << request.toXML();

		conn->request(request, 5000, &rcb, GMSEC_REQUEST_REPUBLISH_NEVER);

		TimeUtil::millisleep(5000);   // allow for the response message(s) to be received

		conn->disconnect();

		test.check("Unexpected response message count", rcb.numRepliesReceived() == cb.expectedResponses());
	}
	catch (const GmsecException& e)
	{
		test.require(e.what(), false);
		return 1;
	}
	return 0;
}

TEST_DRIVER(test_MultiResponse)

File: tests/python3/src/T043_MultiResponse.py
#!/usr/bin/env python3

import libgmsec_python3 as lp
from Test import Test


class T043_Callback(lp.Callback):
    def __init__(self, test):
        super().__init__()
        self.test = test
        self.statusList = []
        self.statusList.append(lp.Message.ResponseStatus_ACKNOWLEDGEMENT)
        self.statusList.append(lp.Message.ResponseStatus_WORKING_KEEP_ALIVE)
        self.statusList.append(lp.Message.ResponseStatus_WORKING_KEEP_ALIVE)
        self.statusList.append(lp.Message.ResponseStatus_WORKING_KEEP_ALIVE)
        self.statusList.append(lp.Message.ResponseStatus_SUCCESSFUL_COMPLETION)

    def on_message(self, conn, msg):
        try:
            reply = conn.get_message_factory().create_message("RESP.DIR")
            reply.set_subject(self.test.get_subject("RESP.DIR"))

            if msg.has_field("DESTINATION-COMPONENT"):
                reply.add_field( lp.StringField("COMPONENT", msg.get_string_value("DESTINATION-COMPONENT")) )
            else:
                reply.add_field( lp.StringField("COMPONENT", "RESPONDER") )

            if msg.has_field("COMPONENT"):
                reply.add_field( lp.StringField("DESTINATION-COMPONENT", msg.get_string_value("COMPONENT")) )
            else:
                reply.add_field( lp.StringField("DESTINATION-COMPONENT", "FILL") )

            if msg.has_field("REQUEST-ID"):
                reply.add_field(msg.get_field("REQUEST-ID"))

            lp.log_verbose("Received Request:\n" + msg.to_xml())

            for status in self.statusList:
                reply.set_field_value("RESPONSE-STATUS", status)

                lp.log_verbose("Issuing Reply:\n" + reply.to_xml())

                conn.reply(msg, reply)

        except Exception as e:
            lp.log_error(str(e))

    def expected_responses(self):
        return len(self.statusList)


class T043_ReplyCallback(lp.ReplyCallback):
    def __init__(self):
        super().__init__()
        self.repliesReceived = 0

    def on_reply(self, conn, request, reply):
        try:
            lp.log_info("Received Reply with RESPONSE-STATUS: " + reply.get_string_value("RESPONSE-STATUS"))
            self.repliesReceived += 1
        except Exception as e:
            lp.log_error(str(e))

    def on_event(self, conn, status, event):
            lp.log_warning("Received unexpected event: " + status.get())

    def num_replies_received(self):
        return self.repliesReceived


class Test_MultiResponse(Test):

    def test(self):
        config = lp.Config(self.get_config())
        config.add_value("mw-multi-resp", "true")

        cb  = T043_Callback(self)
        rcb = T043_ReplyCallback()

        conn = None

        try:
            conn = lp.Connection(config)

            self.set_standard_fields(conn.get_message_factory())

            conn.connect()
            conn.start_auto_dispatch()

            request = conn.get_message_factory().create_message("REQ.DIR")
            request.set_subject(self.get_subject("REQ.DIR"))
            request.add_field( lp.StringField("DIRECTIVE-STRING", "Do something!") )
            request.add_field( lp.StringField("DESTINATION-COMPONENT", "RESPONDER") )
            request.add_field( lp.U16Field("REQUEST-ID", 0) )

            # subscribe to receive the response message(s) (hack!)
            conn.subscribe(self.get_subject("RESP.DIR.+"))

            # subscribe to process the request message
            conn.subscribe(request.get_subject(), cb)

            lp.log_verbose("Issuing Request:\n" + request.to_xml())

            conn.request(request, 5000, rcb, lp.GMSEC_REQUEST_REPUBLISH_NEVER)

            lp.TimeUtil.millisleep(5000);   # allow for the response message(s) to be received

            self.check("Unexpected response message count", rcb.num_replies_received() == cb.expected_responses())

        except Exception as e:
            self.require(str(e), False)

        if conn != None:
            conn.disconnect()
            conn = None


Test.run('Multi-Response', Test_MultiResponse())
